### ISSUE TITLE
False positives within object_usage_linter

### ISSUE THREAD

**Issue by stephenashton-dhsc** [r-lib/lintr#1527]
There are instances where the object_usage_linter seems to return a false "local variable assigned but may not be used" warning.

```
function_name <- function(value_age_) {

  min_age <- 21
  max_age <- 31

  value_age <- case_when(
    value_age_ < min_age ~ min_age,
    value_age_ >= max_age ~ max_age - 1,
    TRUE ~ value_age_
  )

  return(value_age)
}
```

For example, the above code returns the following results:
`Line 3 [object_usage_linter] local variable 'min_age' assigned but may not be used`
`Line 4 [object_usage_linter] local variable 'max_age' assigned but may not be used`

I get a similar issue with the code below, with Lintr flagging the variable "df":
```
function_name <- function() {

  df <- as_tibble(mtcars)

  this_number <- c(1, 2, 3) %>%
    map(~ another_function(df))

  return(this_number)
}
```

**Comment by stephenashton-dhsc** [comment]
This may be related to #1472, but I'm not sure!

**Comment by AshesITR** [comment]
The problem in both your examples is that `case_when()` and that usage of `map()` use non-standard ways of evaluation. In the AST that we analyze, there are just a bunch of (un-evaluated) formulae which happen to contain the same symbols as these local variables.

Anyway, I'd suggest contacting the author of `codetools` because we relay the results of `codetools::checkUsage()` here.

As an alternative, we could consider adding support for formulae similar to what we do with `glue()` literals.

**Comment by MichaelChirico** [comment]
> As an alternative, we could consider adding support for formulae similar to what we do with glue() literals.

I'm' also not sure that's a good idea, because it would hide true positives too:

```r
foo <- function(data, x) {
  y <- x + 1
  lm(y ~ x, data = data)
}
```

`y` is unused, but it's referenced in the formula. However, in the formula, it's being interpreted within `data`, not the function environment, so it's not the same `y`.

There's no way to figure out statically whether it's a case like `case_when()` or a case like `lm()` without maintaining some complicated rules around which formulas are used in simple NSE contexts like `case_when()` vs. typical formula contexts like `lm()`.

I think `# nolint` is appropriate here.
